test/merger/test_general_merger.py

```diff
@@ -169,3 +171,200 @@ def test_no_merge_across_reset() -> None:
     builder2.reset(0)
     expected_qc = builder2.to_circuit()
     modify_circuit_and_check(qc, general_merger.merge_single_qubit_gates, expected_qc)
+
+
+@pytest.mark.parametrize(
+    ("circuit", "expected_result"),
+    [
+        (
+            CircuitBuilder(2).H(0).barrier(0).H(1).barrier(1).H(0).Rx(0, Float(math.pi / 3)).barrier(0).to_circuit(),
+            """version 3.0
+
+qubit[2] q
+
+H q[0]
+H q[1]
+barrier q[0]
+H q[0]
+Rx(1.0471976) q[0]
+barrier q[1]
+barrier q[0]
+""",
+        ),
+        (
+            CircuitBuilder(2).X(0).barrier(0).X(1).barrier(1).CNOT(0, 1).barrier(1).X(1).to_circuit(),
+            """version 3.0
+
+qubit[2] q
+
+X q[0]
+X q[1]
+barrier q[0]
+barrier q[1]
+CNOT q[0], q[1]
+barrier q[1]
+X q[1]
+""",
+        ),
+        (
+            CircuitBuilder(2).X(0).X(1).barrier(0).barrier(1).X(0).to_circuit(),
+            """version 3.0
+
+qubit[2] q
+
+X q[0]
+X q[1]
+barrier q[0]
+barrier q[1]
+X q[0]
+""",
+        ),
+        (
+            CircuitBuilder(4)
+            .H(0)
+            .barrier(0)
+            .H(1)
+            .barrier(1)
+            .H(2)
+            .barrier(2)
+            .H(3)
+            .barrier(3)
+            .CNOT(0, 3)
+            .barrier(0)
+            .barrier(1)
+            .barrier(3)
+            .to_circuit(),
+            """version 3.0
+
+qubit[4] q
+
+H q[0]
+H q[1]
+H q[2]
+H q[3]
+barrier q[0]
+barrier q[1]
+barrier q[2]
+barrier q[3]
+CNOT q[0], q[3]
+barrier q[0]
+barrier q[1]
+barrier q[3]
+""",
+        ),
+    ],
+    ids=[
+        "anonymous_gate",
+        "CNOT_cannot_go_through_a_group_of_linked_barriers",
+        "X_cannot_go_through_a_group_of_linked_barriers",
+        "circuit_with_4_qubits",
+    ],
+)
+def test_rearrange_barriers(circuit: Circuit, expected_result: str) -> None:
+    from opensquirrel.merger.general_merger import rearrange_barriers
+
+    rearrange_barriers(circuit.ir)
+    assert str(circuit) == expected_result
+
+
+@pytest.mark.parametrize(
+    ("circuit", "expected_result"),
+    [
+        (
+            (
+                CircuitBuilder(2)
+                .H(1)
+                .Rz(1, Float(2.332))
+                .H(0)
+                .Rx(0, Float(1.423))
+                .barrier(0)
+                .barrier(1)
+                .H(1)
+                .Rz(1, Float(2.332))
+                .H(0)
+                .Rx(1, Float(1.423))
+                .barrier(0)
+                .barrier(1)
+                .to_circuit()
+            ),
+            """version 3.0
+
+qubit[2] q
+
+Anonymous gate: BlochSphereRotation(Qubit[0], axis=[ 0.60376 -0.52053  0.60376], angle=-2.18173, phase=1.5708)
+Anonymous gate: BlochSphereRotation(Qubit[1], axis=[0.36644 0.85525 0.36644], angle=-1.72653, phase=1.5708)
+barrier q[0]
+barrier q[1]
+H q[0]
+Anonymous gate: BlochSphereRotation(Qubit[1], axis=[ 0.28903 -0.42028 -0.86013], angle=1.66208, phase=1.5708)
```

Comment:
Is this correct?

```
.H(1)
.Rz(1, Float(2.332))
.H(0)
.Rx(1, Float(1.423))
```

gets merged into?

```
H q[0]
Anonymous gate: BlochSphereRotation(Qubit[1], axis=[ 0.28903 -0.42028 -0.86013], angle=1.66208, phase=1.5708)
```

Reply:
If I understand correctly, you mean to say that it may not be correct because the placement of H(0) is placed after the H(1) and the output has the inverse? 

If this is the case, then yes it is correct because we only care about the positioning of barriers, not gates in general.

Reply:
OK, so `merge_single_qubit_gates` is accumulating here 2 Bloch sphere rotations, 1 for `q[0]`, and 1 for `q[1]`. When the first barrier comes, it spits both BSR out. So my question is if:

```
.H(1) 
.Rz(1, Float(2.332)) 
.Rx(1, Float(1.423))
```

is equivalent to:

```
Anonymous gate: BlochSphereRotation(Qubit[1], axis=[ 0.28903 -0.42028 -0.86013], angle=1.66208, phase=1.5708)
```

Reply:
![image](https://github.com/user-attachments/assets/247c22d2-6a70-4a1f-8a6b-77c27195a090)

![image](https://github.com/user-attachments/assets/d77083a0-8c68-4f6e-b396-f80fdf86222f)

Yes, it is correct, off by a global phase.